inc/post-types/timeline.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+/*
+ * Open Development
+ * timeline
+ */
+
+class Odm_Timeline {
+
+	function __construct() {
+
+		add_action('init', array($this, 'register_post_type'));
+		add_action('add_meta_boxes', array($this, 'add_meta_box'));
+		add_action('save_post', array($this, 'save_post_data'));
+
+	}
+
+	function register_post_type() {
+
+		$labels = array(
+			'name'               => _x( 'Timelines', 'post type general name', 'odm' ),
+			'singular_name'      => _x( 'Timeline', 'post type singular name', 'odm' ),
+			'menu_name'          => _x( 'Timelines', 'admin menu', 'odm' ),
+			'name_admin_bar'     => _x( 'Timeline', 'add new on admin bar', 'odm' ),
+			'add_new'            => _x( 'Add new', 'timeline', 'odm' ),
+			'add_new_item'       => __( 'Add new timeline', 'odm' ),
+			'new_item'           => __( 'New timeline', 'odm' ),
+			'edit_item'          => __( 'Edit timeline', 'odm' ),
+			'view_item'          => __( 'View timeline', 'odm' ),
+			'all_items'          => __( 'All timelines', 'odm' ),
+			'search_items'       => __( 'Search timelines', 'odm' ),
+			'parent_item_colon'  => __( 'Parent timelines:', 'odm' ),
+			'not_found'          => __( 'No timelines found.', 'odm' ),
+			'not_found_in_trash' => __( 'No timelines found in trash.', 'odm' )
+		);
+
+		$args = array(
+			'labels'             => $labels,
+			'public'             => true,
+			'publicly_queryable' => true,
+			'show_ui'            => true,
+			'show_in_menu'       => true,
+			'menu_icon' 				 => 'dashicons-calendar-alt',
+			'query_var'          => true,
+			'rewrite'            => array( 'slug' => 'timelines' ),
+			'capability_type'    => 'post',
+			'has_archive'        => true,
+			'hierarchical'       => false,
+			'menu_position'      => 5,
+			'taxonomies'         => array( 'category', 'post_tag'),
+			'supports'           => array( 'title', 'editor', 'author', 'thumbnail', 'excerpt', 'revisions')
+		);
+
+		register_post_type( 'timeline', $args );
+
+	}
+
+
+	function add_meta_box($post_type) {
+		$list_post_types = array('profiles', 'topic');
+		if(in_array($post_type, $list_post_types)){
+			add_meta_box(
+					'related-timeline',
+					__('Add Related Timeline', 'odm'),
+					array($this, 'add_related_timeline_box'),
+					$post_type,
+					'advanced',
+					'high'
+			);
+		}
+	}
+
+	function add_related_timeline_box($post = false){
+		$related_timeline_post = get_post_meta($post->ID, '_related_timeline_post', true);
```

Comment:
@huyeng the scenario that you are proposing can be achieved also with the mechanisms explained on https://docs.google.com/a/ewmi-odi.org/document/d/11zXdHh9mzbuH5bEvqNCh62Iz12ukLU0JeXy2KxtrMeQ/edit?usp=drive_web

See the **representation template** option on the widget configuration. if we define a template **show contents html** editors would be able to choose it so the HTML content of the timeline gets embedded as you are expecting.

please do not get confused by the use case UC2 on that document. as opposed to the use case you are presenting in your comments here, that one generates the timeline automatically from CKAN content.

the important point I want to raise, is that we should not have 2 functions that can do the same thing. Otherwise the processes get too messy